File: packages/picoipc/python/examples/stress.py
#!/usr/bin/env python3
"""Fill the ring until full; verify backpressure and unblock after consume."""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from smh_q import Ring


def main() -> int:
    name = f"smh_q_py_stress_{os.getpid()}"
    Ring.unlink(name)

    slot_count = 4
    ring = Ring(name, create=True, slot_count=slot_count, slot_size=128)

    msg = b"x"
    published = 0
    while ring.try_publish(msg):
        published += 1

    print(f"filled ring: published={published} slot_count={slot_count} (ring full)")
    assert published == slot_count, f"expected {slot_count} publishes, got {published}"
    assert not ring.try_publish(msg), "try_publish should fail when ring is full"

    got = ring.try_consume()
    assert got == msg, f"expected {msg!r}, got {got!r}"

    assert ring.try_publish(msg), "try_publish should succeed after consume"

    print(f"backpressure OK: full at {slot_count} slots, unblock after consume")

    ring.close(unlink=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/picoipc/examples/cpp/stress.cpp
#include "smh_q/ring.hpp"

#include <cstdio>
#include <cstring>
#include <string>
#include <unistd.h>

#define CHECK(cond)                                                          \
  do {                                                                       \
    if (!(cond)) {                                                           \
      std::fprintf(stderr, "CHECK failed: %s (%s:%d)\n", #cond, __FILE__,   \
                   __LINE__);                                                \
      std::abort();                                                          \
    }                                                                        \
  } while (0)

int main() {
  const std::string name = "smh_q_stress_" + std::to_string(::getpid());
  smh_q::Ring::unlink(name);

  constexpr std::uint32_t kSlots = 4;
  smh_q::Ring::Config cfg;
  cfg.name = name;
  cfg.slot_count = kSlots;
  cfg.slot_size = 128;

  smh_q::Ring ring(cfg, true);

  const char* msg = "x";
  const auto span = std::span<const std::byte>(reinterpret_cast<const std::byte*>(msg), 1);

  int published = 0;
  while (ring.try_publish(span)) {
    ++published;
  }
  std::printf("filled ring: published=%d slot_count=%u (ring full)\n", published, kSlots);
  CHECK(published == static_cast<int>(kSlots));

  CHECK(!ring.try_publish(span));

  auto got = ring.try_consume();
  CHECK(got.has_value());
  CHECK(got->size() == 1);
  CHECK(got->front() == std::byte{'x'});

  CHECK(ring.try_publish(span));

  std::printf("backpressure OK: full at %d slots, unblock after consume\n", kSlots);

  smh_q::Ring::unlink(name);
  return 0;
}
